File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/ahn-product-1-0/IndexRef.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class IndexRef : ISerialized
    {
        public IndexRef(XmlNode xmlNode)
        {
            XmlNode underlyingIndexRefNode = xmlNode.SelectSingleNode("underlyingIndexRef");
            
            if (underlyingIndexRefNode != null)
            {
                if (underlyingIndexRefNode.Attributes["href"] != null || underlyingIndexRefNode.Attributes["id"] != null) 
                {
                    if (underlyingIndexRefNode.Attributes["id"] != null) 
                    {
                        underlyingIndexRefIDRef_ = underlyingIndexRefNode.Attributes["id"].Value;
                        UnderlyingIndex ob = new UnderlyingIndex(underlyingIndexRefNode);
                        IDManager.SetID(underlyingIndexRefIDRef_, ob);
                    }
                    else if (underlyingIndexRefNode.Attributes["href"] != null)
                    {
                        underlyingIndexRefIDRef_ = underlyingIndexRefNode.Attributes["href"].Value;
                    }
                    else
                    {
                        underlyingIndexRef_ = new UnderlyingIndex(underlyingIndexRefNode);
                    }
                }
                else
                {
                    underlyingIndexRef_ = new UnderlyingIndex(underlyingIndexRefNode);
                }
            }
            
        
            XmlNode eventDateRefNode = xmlNode.SelectSingleNode("eventDateRef");
            
            if (eventDateRefNode != null)
            {
                if (eventDateRefNode.Attributes["href"] != null || eventDateRefNode.Attributes["id"] != null) 
                {
                    if (eventDateRefNode.Attributes["id"] != null) 
                    {
                        eventDateRefIDRef_ = eventDateRefNode.Attributes["id"].Value;
                        EventDate ob = new EventDate(eventDateRefNode);
                        IDManager.SetID(eventDateRefIDRef_, ob);
                    }
                    else if (eventDateRefNode.Attributes["href"] != null)
                    {
                        eventDateRefIDRef_ = eventDateRefNode.Attributes["href"].Value;
                    }
                    else
                    {
                        eventDateRef_ = new EventDate(eventDateRefNode);
                    }
                }
                else
                {
                    eventDateRef_ = new EventDate(eventDateRefNode);
                }
            }
            
        
            XmlNode fixingDaysNode = xmlNode.SelectSingleNode("fixingDays");
            
            if (fixingDaysNode != null)
            {
                if (fixingDaysNode.Attributes["href"] != null || fixingDaysNode.Attributes["id"] != null) 
                {
                    if (fixingDaysNode.Attributes["id"] != null) 
                    {
                        fixingDaysIDRef_ = fixingDaysNode.Attributes["id"].Value;
                        FixingDays ob = new FixingDays(fixingDaysNode);
                        IDManager.SetID(fixingDaysIDRef_, ob);
                    }
                    else if (fixingDaysNode.Attributes["href"] != null)
                    {
                        fixingDaysIDRef_ = fixingDaysNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fixingDays_ = new FixingDays(fixingDaysNode);
                    }
                }
                else
                {
                    fixingDays_ = new FixingDays(fixingDaysNode);
                }
            }
            
        
        }
        
    
        #region UnderlyingIndexRef_
        private UnderlyingIndex underlyingIndexRef_;
        public UnderlyingIndex UnderlyingIndexRef_
        {
            get
            {
                if (this.underlyingIndexRef_ != null)
                {
                    return this.underlyingIndexRef_; 
                }
                else if (this.underlyingIndexRefIDRef_ != null)
                {
                    underlyingIndexRef_ = IDManager.getID(underlyingIndexRefIDRef_) as UnderlyingIndex;
                    return this.underlyingIndexRef_; 
                }
                else
                {
                      return this.underlyingIndexRef_; 
                }
            }
            set
            {
                if (this.underlyingIndexRef_ != value)
                {
                    this.underlyingIndexRef_ = value;
                }
            }
        }
        #endregion
        
        public string underlyingIndexRefIDRef_ { get; set; }
        #region EventDateRef_
        private EventDate eventDateRef_;
        public EventDate EventDateRef_
        {
            get
            {
                if (this.eventDateRef_ != null)
                {
                    return this.eventDateRef_; 
                }
                else if (this.eventDateRefIDRef_ != null)
                {
                    eventDateRef_ = IDManager.getID(eventDateRefIDRef_) as EventDate;
                    return this.eventDateRef_; 
                }
                else
                {
                      return this.eventDateRef_; 
                }
            }
            set
            {
                if (this.eventDateRef_ != value)
                {
                    this.eventDateRef_ = value;
                }
            }
        }
        #endregion
        
        public string eventDateRefIDRef_ { get; set; }
        #region FixingDays_
        private FixingDays fixingDays_;
        public FixingDays FixingDays_
        {
            get
            {
                if (this.fixingDays_ != null)
                {
                    return this.fixingDays_; 
                }
                else if (this.fixingDaysIDRef_ != null)
                {
                    fixingDays_ = IDManager.getID(fixingDaysIDRef_) as FixingDays;
                    return this.fixingDays_; 
                }
                else
                {
                      return this.fixingDays_; 
                }
            }
            set
            {
                if (this.fixingDays_ != value)
                {
                    this.fixingDays_ = value;
                }
            }
        }
        #endregion
        
        public string fixingDaysIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/ahn-product-1-0/IndexRef.cpp
// IndexRef.cpp 
#include "IndexRef.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

IndexRef::IndexRef(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //underlyingIndexRefNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* underlyingIndexRefNode = xmlNode->FirstChildElement("underlyingIndexRef");

   if(underlyingIndexRefNode){underlyingIndexRefIsNull_ = false;}
   else{underlyingIndexRefIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- underlyingIndexRefNode , address : " << underlyingIndexRefNode << std::endl;
   #endif
   if(underlyingIndexRefNode)
   {
      if(underlyingIndexRefNode->Attribute("href") || underlyingIndexRefNode->Attribute("id"))
      {
          if(underlyingIndexRefNode->Attribute("id"))
          {
             underlyingIndexRefIDRef_ = underlyingIndexRefNode->Attribute("id");
             underlyingIndexRef_ = boost::shared_ptr<UnderlyingIndex>(new UnderlyingIndex(underlyingIndexRefNode));
             underlyingIndexRef_->setID(underlyingIndexRefIDRef_);
             IDManager::instance().setID(underlyingIndexRefIDRef_,underlyingIndexRef_);
          }
          else if(underlyingIndexRefNode->Attribute("href")) { underlyingIndexRefIDRef_ = underlyingIndexRefNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { underlyingIndexRef_ = boost::shared_ptr<UnderlyingIndex>(new UnderlyingIndex(underlyingIndexRefNode));}
   }

   //eventDateRefNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* eventDateRefNode = xmlNode->FirstChildElement("eventDateRef");

   if(eventDateRefNode){eventDateRefIsNull_ = false;}
   else{eventDateRefIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- eventDateRefNode , address : " << eventDateRefNode << std::endl;
   #endif
   if(eventDateRefNode)
   {
      if(eventDateRefNode->Attribute("href") || eventDateRefNode->Attribute("id"))
      {
          if(eventDateRefNode->Attribute("id"))
          {
             eventDateRefIDRef_ = eventDateRefNode->Attribute("id");
             eventDateRef_ = boost::shared_ptr<EventDate>(new EventDate(eventDateRefNode));
             eventDateRef_->setID(eventDateRefIDRef_);
             IDManager::instance().setID(eventDateRefIDRef_,eventDateRef_);
          }
          else if(eventDateRefNode->Attribute("href")) { eventDateRefIDRef_ = eventDateRefNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { eventDateRef_ = boost::shared_ptr<EventDate>(new EventDate(eventDateRefNode));}
   }

   //fixingDaysNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fixingDaysNode = xmlNode->FirstChildElement("fixingDays");

   if(fixingDaysNode){fixingDaysIsNull_ = false;}
   else{fixingDaysIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fixingDaysNode , address : " << fixingDaysNode << std::endl;
   #endif
   if(fixingDaysNode)
   {
      if(fixingDaysNode->Attribute("href") || fixingDaysNode->Attribute("id"))
      {
          if(fixingDaysNode->Attribute("id"))
          {
             fixingDaysIDRef_ = fixingDaysNode->Attribute("id");
             fixingDays_ = boost::shared_ptr<FixingDays>(new FixingDays(fixingDaysNode));
             fixingDays_->setID(fixingDaysIDRef_);
             IDManager::instance().setID(fixingDaysIDRef_,fixingDays_);
          }
          else if(fixingDaysNode->Attribute("href")) { fixingDaysIDRef_ = fixingDaysNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fixingDays_ = boost::shared_ptr<FixingDays>(new FixingDays(fixingDaysNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<UnderlyingIndex> IndexRef::getUnderlyingIndexRef()
{
   if(!this->underlyingIndexRefIsNull_){
        if(underlyingIndexRefIDRef_ != ""){
             return boost::shared_static_cast<UnderlyingIndex>(IDManager::instance().getID(underlyingIndexRefIDRef_));
        }else{
             return this->underlyingIndexRef_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<UnderlyingIndex>();
   }
}
boost::shared_ptr<EventDate> IndexRef::getEventDateRef()
{
   if(!this->eventDateRefIsNull_){
        if(eventDateRefIDRef_ != ""){
             return boost::shared_static_cast<EventDate>(IDManager::instance().getID(eventDateRefIDRef_));
        }else{
             return this->eventDateRef_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EventDate>();
   }
}
boost::shared_ptr<FixingDays> IndexRef::getFixingDays()
{
   if(!this->fixingDaysIsNull_){
        if(fixingDaysIDRef_ != ""){
             return boost::shared_static_cast<FixingDays>(IDManager::instance().getID(fixingDaysIDRef_));
        }else{
             return this->fixingDays_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<FixingDays>();
   }
}
}

